### ISSUE TITLE
Android Auto launcher not showing sideloaded OsmAnd for Android 11 and above 

### ISSUE THREAD

**Issue by polarbearing** [osmandapp/OsmAnd#15400]
There are several discussions where OsmAnd (plus or not) does not show in the Android Auto launcher when not installed directly from G. Playstore, i.e. installed from other stores (Fdroid, Aurora, Amazon) or sideloaded. 
The behaviour and workaround depend on the Android version, but in Android 11 and above it differs between VLC and OsmAnd. 
I compared an Android 7 and an Android 11 device. 
My problem is that currently I cannot test nightly builds of OsmAnd under AA in Android 11, and AA itself stops working in Android 7.
____
***Android 7*** with the last AA version that works under Android 7
Install without package installer name:
```
adb push VLC_for_Android_v3.3.0_armeabi-v7a.apk /data/local/tmp/tmp.apk
adb shell
$ pm install /data/local/tmp/tmp.apk                                                               
Success
```
Now the VLC is listed in AA, but does ***not*** show in the launcher
(screenshots follow)
We simply reinstall the same file and set the package installer name:
```
$ pm install -r -i "com.android.vending" /data/local/tmp/tmp.apk                                   
Success
```
Now the VLC also shows in the Launcher. 
The same way, OsmAnd used to work until summer 2022 in AA. It still shows in the Launcher.
Now AA either throws timebombs, declaring things are too old, or throws communication errors claiming the clocks between car and phone were allegedly out of sync (which they are not)
____
In ***Android 11***, apparently the ***package installer name*** is not needed anymore.
* AA version 8.0.623354, where AA has no app on the phone anymore, but a car screen only, 
* The VLC is the latest downloaded from the videolan.org website. We explicitly set the installer name to nonsense, 
* An Osmand nightly was installed previously with setting the playstore installer name:
```
adb push VLC-Android-3.5.2-arm64-v8a.apk /data/local/tmp/tmp.apk
adb shell
$ pm install  -i "bla.bla.bla" /data/local/tmp/temp.apk 
$ pm list packages -i |grep org 
package:org.videolan.vlc  installer=null
$ pm list packages -i |grep osm
package:net.osmand.plus  installer=com.android.vending
```
Now, while both VLC and OsmAnd+ are listed in the AA settings, only VLC is shown in the Launcher. Apparently, the installer name does not matter anymore. 

AA *developer setting "Unknown sources"* is ticked. When I untick that, the VLC remains listed in AA, but disappears from the Launcher. I tick it again, it is back. Not so for OsmAnd. 
(screenshots follow)

While it is known that VLC and OsmAnd are in different categories, media and navigation, I see no reason why Google would it make easier for media apps to be sideloaded than for navigation apps. 

Related Discussions:
#13514
#3391
[Fdroid: Osmand~: 4.1.9 doesn't show up in Android Auto](https://gitlab.com/fdroid/fdroiddata/-/issues/2567)
[microG + Android Auto fully working](https://forum.xda-developers.com/t/microg-android-auto-fully-working.4319159/page-6)

**Comment by vshcherb** [comment]
That's quite unfortunate cause now we're stuck to Beta-Release cycle on Google Play

**Comment by polarbearing** [comment]
Unless we find the reason why a sideloaded VLC behaves different from a sideloaded OsmAnd.

**Comment by vshcherb** [comment]
You mentionned it's a media apps there were less strict and different perspective from the beginning. FOR  a long time only waze and google was navigation only but vlc was already there

**Comment by kevinmcmurtrie** [comment]
Does the Play Store version support microSD card maps yet?  I'm on F-Droid builds solely to avoid Google's monopolistic permissions rules.

**Comment by DaveIsC** [comment]
Hello, maybe I found a solution hoe to display sideloaded OsmAnd on android auto launcher. For me, this solution works (I have Android 13 and the latest official public version of Android auto. 

If this is not the case of this issue, I'm sorry :-D

First you need to download the F-droid application (from here https://f-droid.org/). From there, download and install OsmAnd~ 4.2.7.apk (the app should contain ~ in the name). Then download any app from Google play (it doesn't matter what kind of app) like "apk extractor". Export the installed OsmAnd app and then uninstall OsmAnd. Then download and install King Installer 1.2 app (for example from here https:/ /oceanofapk.com/king-installer-v1-2-no-root-altered-apk-free-download/) (not found on Google play) Open King Installer and go to the folder where the exported app OsmAnd~ 4.2.7 is located. Now install it using King Installer. Then open the newly installed OsmAnd~ app and download the maps you need. Go to your car, connect your phone and the app should show up (just to be sure, I turned on developer options and enabled unknown sources) .It works for me.

One thing I noticed is that this app version 4.2.7 and specifically from F-droid doesn't have a "purchases or restore purchases" item in the settings. The original version 4.2.7 has it but this one doesn't. This version is probably the only one that works with android auto (another version I tried to install with King installer, got an error message and could not launch OsmAnd app. With King Installer you need to install all apps when you want to sideload them. It also worked for me on Sygic.

**Comment by michaelblyons** [comment]
## Root cause

The actual root cause is unclear, but [Google is pretty explicit](https://developer.android.com/docs/quality-guidelines/car-app-quality) about what users should expect in Android Auto:

> &bigstar; **Important:** Apps must meet the criteria listed in this page to qualify as an Android Auto or Android Automotive OS app on Google Play. Apps don't work with Android Auto unless they are installed from the Play Store or the ONE store.

Clearly this is not strictly followed for some apps, since people have had success sideloading messaging and media apps.

## Workarounds

So far, I've only seen two major types of workaround.

### Lie about installer source

Since Google only allows this app when "installed from the Play Store," pretend that it is. There are a couple ways to do this:

- [Use `adb`](https://gitlab.com/fdroid/fdroiddata/-/issues/2567#note_897966417)

    ```sh
     adb push net.osmand.plus_421.apk /data/local/tmp
     adb shell pm install -i "com.android.vending" -r /data/local/tmp/net.osmand.plus_421.apk
    ```

- [Use an installer app](https://gitlab.com/fdroid/fdroiddata/-/issues/2567#note_1209405095) that [claims to be the Play Store](https://gitlab.com/fdroid/fdroiddata/-/issues/2567#note_1209450609).

### Patch Android Auto to allow your app

If your phone is rooted and XPrivacy is installed, @UnlockAA has [a module](https://github.com/UnlockAA/XPrivacyLuaHook) that will do this.

**Comment by starbrights** [comment]
If I remember right this adb install trick works only on <=A11, right?

**Comment by flo-at** [comment]
> If I remember right this adb install trick works only on <=A11, right?

I'm using Android 12 and I used adb to spoof the installer. In the app's details it sais that it was installed from the Play Store. It doesn't show up in Android Auto though. So I guess you're right, unfortunately. :disappointed:

My phone is not rooted so I guess there aren't many options.

**Comment by Thestroyer** [comment]
Yeah, unfortunately it does not work in higher Android versions (Android 14). It's interesting though, it shows "installed from the Play Store" and still does not authorize it, how is it authorized then?

**Comment by flo-at** [comment]
Google recently changed the policy (https://developer.android.com/docs/quality-guidelines/car-app-quality). Now they don't mention that Android Auto only works for apps installed from the Play Store anymore. I cannot test it because I'm stuck at an older version of Android. If they still don't allow this, it probably violates the Digital Markets Act for installations in the EU.

**Comment by Thestroyer** [comment]
Yeah AA works for some other apps (from other sources), but you need to enable developer mode...unfortunately maps are different and need to be installed from play store

**Comment by souija-wtf** [comment]
Looks like there is a difference between running on the SDK desktop-head-unit and a real car also. DHU is working with a nightly build perfectly well, but a real head unit isn't showing it at all, not even in the launcher config.

Running on GrapheneOS / Pixel 8 Pro / Sandboxed Google Play / APK installed OsmAnd. Not exactly standard, but something has changed recently on the AA side of things which breaks OsmAnd in this setup.

**Comment by khimaros** [comment]
i wonder if it is possible to launch an intent with `adb`, this might be an acceptable workaround for the lack of launcher icon.

**Comment by starbrights** [comment]
I thought that one should help: 
https://github.com/sn-00-x/aa4mg
But I seems that I don't really understand that section 3rd Party with this LSPosed.
I installed that two Magisk modules, but how to proceed? Should there appear an LSPosed app by installing Magisk Modules (zip)?

**Comment by M0P** [comment]
Hi everyone! I just wanted to share that I was able to get this to work by following this comment: https://gitlab.com/fdroid/fdroiddata/-/issues/2567#note_2042922228.

It uses https://github.com/chenxiaolong/AlterInstaller, and you'll need at least temporary root. I hope this helps!

**Comment by bill-ba** [comment]
> Hi everyone! I just wanted to share that I was able to get this to work by following this comment: https://gitlab.com/fdroid/fdroiddata/-/issues/2567#note_2042922228.
> 
> It uses https://github.com/chenxiaolong/AlterInstaller, and you'll need at least temporary root. I hope this helps!

+1
Tested and working with Los21/microG  + OSMand~ via FDroid :
- install Magisk
- install aa4mg via Magisk in order to run Android Auto
- install Osmand via AuroraStore + 'Root installer' or 'Aurora Services' as the installation method
- install AlterInstaller via Magisk
- create the file /data/local/tmp/AlterInstaller.json, code example :
```
{
  // The top level key is the package to modify.
  // The values below can be omitted or set to null
  // to leave the fields unchanged.
  "net.osmand.plus": {
    // Mark OSM as being installed by the Play Store.
	"installer": "com.android.vending",
	// Mark OSM as being only updatable by Droid-ify.
	//"updateOwner": "com.looker.droidify"
	// Mark OSM as being only updatable by F-Droid
	"updateOwner": "org.fdroid.fdroid"
  }
}
```
- last reboot & enjoy :)
